storage/statedb/hasher.go

```diff
@@ -232,10 +251,56 @@ func (h *hasher) store(n node, db *Database, force bool) (node, uint16) {
 	return hash, lenEncoded
 }
 
-func (h *hasher) makeHashNode(data []byte) hashNode {
-	n := make(hashNode, h.sha.Size())
+func (h *hasher) makeHashNode(data []byte, onRoot bool) hashNode {
+	var hash common.Hash
 	h.sha.Reset()
 	h.sha.Write(data)
-	h.sha.Read(n)
-	return n
+	h.sha.Read(hash[:])
+	if h.pruning && (h.storageRoot || !onRoot) {
+		return hash.Extend().Bytes()
+	} else {
+		return hash.ExtendLegacy().Bytes()
+	}
+}
+
+func (h *hasher) nodeForHashing(original node) node {
+	return unextendNode(original, false)
+}
+
+func (h *hasher) nodeForStoring(original node) node {
+	return unextendNode(original, true)
+}
+
+func unextendNode(original node, preserveExtHash bool) node {
```

Comment:
Could you explain the meaning of unextending a node while preserving ext hash?

Reply:
Even if `preserveExtHash==false`, legacy nonces are stripped anyway.
- case hashNode has the conditional in this function.
- case valueNode has the conditional inside SCA.EncodeRLPExt.

Reply:
I'm sorry, I meant to ask when `preserveExtHash` is true.

Reply:
Oh, same here.. I described when `preserveExtHash==true` in the above comment.
- when preserveExtHash==true, only strips legacy nonces. Used before storing the node in database.
- when preserveExtHash==false, strips any nonces. Used before calculating the Merkle hash.

Reply:
gotcha, thanks